### ISSUE TITLE
Storm MaterialX shader compilation errors on Metal

### ISSUE THREAD

**Issue by pablode** [PixarAnimationStudios/USD#2479]
### Description of Issue

Trying to render a simple USD+MaterialX file on MacOS produces following MSL compilation errors:

```
--- tmp/USD ‹dev* M?› » usdview ~/tmp/usd-assets/cornell-box/mtlx-usdshade/cornell-box.usda                                                           1 ↵
Warning: in _ValidateCompilation at line 228 of /Users/pablode/tmp/USD/pxr/imaging/hdSt/glslProgram.cpp -- Failed to compile shader (FRAGMENT_SHADER): Compilation failed:

program_source:1388:8: error: redefinition of 'BSDF'
struct BSDF { float3 response; float3 throughput; float thickness; float ior; };
       ^
program_source:1144:8: note: previous definition is here
struct BSDF { float3 response; float3 throughput; float thickness; float ior; };
       ^
program_source:1390:8: error: redefinition of 'surfaceshader'
struct surfaceshader { float3 color; float3 transparency; };
       ^
program_source:1146:8: note: previous definition is here
struct surfaceshader { float3 color; float3 transparency; };
       ^
program_source:1391:8: error: redefinition of 'volumeshader'
struct volumeshader { float3 color; float3 transparency; };
       ^
program_source:1147:8: note: previous definition is here
struct volumeshader { float3 color; float3 transparency; };
       ^
program_source:1392:8: error: redefinition of 'displacementshader'
struct displacementshader { float3 offset; float scale; };
       ^
program_source:1148:8: note: previous definition is here
struct displacementshader { float3 offset; float scale; };
       ^
program_source:1393:8: error: redefinition of 'lightshader'
struct lightshader { float3 intensity; float3 direction; };
       ^
program_source:1149:8: note: previous definition is here
struct lightshader { float3 intensity; float3 direction; };
       ^
```

The problem is that `emitTypeDefinitions()` is called two times: in [_EmitConstantsUniformsAndTypeDefs](https://github.com/PixarAnimationStudios/USD/blob/96f9ec81f264f41c28e1bb169a49ea6f1ac483f2/pxr/imaging/hdSt/materialXShaderGen.cpp#L690) and in [_EmitGlslfxMetalShader](https://github.com/PixarAnimationStudios/USD/blob/96f9ec81f264f41c28e1bb169a49ea6f1ac483f2/pxr/imaging/hdSt/materialXShaderGen.cpp#L1108). Getting rid of the former resolves the error.

### Steps to Reproduce
1. Open the mtlx-usdshade Cornell Box from [here](https://github.com/pablode/usd-assets/tree/main/cornell-box) in usdview

### System Information (OS, Hardware)
macOS Catalina 10.15.7
Intel Iris Plus Graphics 650

### Package Versions
Latest dev branch (96f9ec8)
MaterialX 1.38.7

### Build Flags
`--materialx --opencolorio --openimageio`

**Comment by pablode** [comment]
Ah, I see this has been fixed in 3d34a55. Unfortunately, by removing the latter occurrence, another error is introduced.

For the [veach-ajar](https://github.com/pablode/usd-assets/tree/main/veach-ajar) scene, the following is thrown:
```
--- tmp/USD ‹dev* M?› » usdview ~/tmp/usd-assets/veach-ajar/veach-ajar.usdc                                                                                                             130 ↵
Warning: in _ValidateCompilation at line 228 of /Users/pablode/tmp/USD/pxr/imaging/hdSt/glslProgram.cpp -- Failed to compile shader (FRAGMENT_SHADER): Compilation failed:

program_source:1057:5: error: unknown type name 'EDF'
    EDF edf;
    ^
```

**Comment by jesschimein** [comment]
Filed as internal issue #USD-8409
